---
title: "Targeted Selection"
author: "Dom Grisafe"
date: "`r Sys.Date()`"
output:
  html_document:
    keep_md: FALSE
    toc: TRUE
    toc_float: TRUE
    toc_depth: 2
csl: hepatology.csl
bibliography: "/Users/domusc/Dropbox/Zotero_Library/!ZoteroLibrary.bib"
---

```{r setup_select, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(styledom)
library(kableExtra)
library(cowplot)
  theme_set(theme_cowplot())
library(circlize)

source("!load_clean_data.R")
```

# Introduction

This analysis aims to identify features that relate to liver injury genetic SNPs and liver injury enzymes using **multivariable linear regression**, and then identify **partial correlations** of these targeted features with all available features. The results are then visualized using **circos globe plots**. Finally, features that are significantly identified in at least 2 targeted analyses will be considered **enriched** and studied for further analysis. All linear regressions and correlations are adjusted for the same set of covariates, which are discussed below.

This analysis has four parts:

0) **Review Outcomes and Covariates** that will be used for regression and correlation analysis

1) Identify **Target Features** for each liver injury outcome within each omics layer, adjusting for covariates. This will yield 5 lists of targeted features for each outcome:

    - *Primary Outcome* — target features associated with NAFLD-specific liver injury defined by sex-specific cutoffs for ALT [@schwimmer_2010_safety].
    
    - *Secondary Outcome* — target features associated with general liver injury defined by any liver enzyme (ALT, AST, GGT) elevated ≥ 90th percentile within HELIX [@stratakis_2020_prenatal].
    
    - *Each Liver Enzyme* separately as a continuous measure

2) Identify significant **Partial Correlations** between the targeted omics features and all available features from the four remaining omics layers after adjusting for covariates.

3) Visualize partial correlations for each omics layer using **Circos Globe Plots** [@gu_2014_circlize].

4) Identify **Enriched Omics Features** that were significantly correlated with targeted features in two or more correlation analyses.

## Alpha Level
```{r}
alpha_pval_beta <- 0.05
```

The alpha level of significance is `r alpha_pval_beta` for linear regression, partial correlation, and multiple comparison adjustment by false discovery rate.



# Review Outcomes and Covariates

## Outcomes

Liver injury was defined as:

1) **Primary Outcome** — NAFLD liver injury defined as ALT ≥ 22.1 U/L for girls and ≥ 25.8 U/L for boys [@schwimmer_2010_safety]. 

2) **Secondary Outcome** — Any liver enzyme (ALT, AST, or GGT) ≥ 90th percentile within HELIX cohort, which has been used previously in HELIX [@stratakis_2020_prenatal].  

## Covariates

Discussion with lab group 12/15/2020, only adjust for cohort, age, and sex of children.

We considered adjusting for **child modifiable risk factors** and **maternal sociodemographics**, but decided they did not meet the definition of a confounder and therefore did not include them in multivariable models.

<!-- Covariates were selected in a four-step process:   -->

<!-- 1) Discussion with NS on 11/20/2020 to include **child sociodemographics**, **child modifiable risk factors**, and **parental sociodemographics**.   -->

<!-- 2) Exploratory Data Analysis identified several colinear variables including sugar consumption and cohort, and paternal and maternal age and education.   -->

<!-- 3) Remove diet covariates as they likely over-adjust for relationship between genes and multi-omics layers.   -->

<!-- 4) Prefer categorized BMI variables to continuous because more applicable to clinical decision making -->

```{r}
# covariates
names_covars <- c(
  
  # Child sociodemographics
  "h_cohort", "hs_child_age_days_None", "e3_sex_None" 
  
  # after lab discussion 12/15/2020, remove all covariates potentially intermediates
  # "h_ethnicity_c_None", "e3_bwc_None", 
  
  # Child Modifiable Risk Factors
  # "hs_bmicat_None",
  # colinear: "h_sugar_post_Log" and "h_cohort"
  # over-adjustment: "h_sugar_post_Log" and "h_nonalc_post_Log"
  # linear BMI: "hs_c_bmi_None"
  
  # Parental sociodemographics
  # "h_age_None", "h_edumc_None", "h_mbmic_None", "hs_wgtgain_None"
  # colinear: "h_age_None" and "h_fage_None", "h_edufc_None" and "h_edumc_None"
  # linear BMI: hs_c_bmi_None
)
```

## General Linear Model

We will adjust all linear models for the same covariates

$$
\begin{eqnarray}
Y = \alpha + \beta_1X
     &+& \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
     \\
     \\
Y = \alpha + \beta_1X
     &+& \color{red}{\beta_2X_{2\text{ Cohort}}}
     + \color{red}{\beta_3X_{3\text{ Age}}}
     + \color{red}{\beta_4X_{4\text{ Sex}}}
\end{eqnarray}
$$
<!-- $$ -->
<!-- \begin{eqnarray} -->
<!-- Y = \alpha + \beta_1X -->
<!--      &+& \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}   -->
<!--      + \color{purple}{\beta_jX_{j\text{ Child Modifiable Risk Factors}}}   -->
<!--      + \color{blue}{\beta_kX_{k\text{ Maternal Sociodemographics}}} -->
<!--      \\ -->
<!--      \\ -->
<!-- Y = \alpha + \beta_1X -->
<!--      &+& \color{red}{\beta_2X_{2\text{ Cohort}}} -->
<!--      + \color{red}{\beta_3X_{3\text{ Age}}} -->
<!--      + \color{red}{\beta_4X_{4\text{ Sex}}} -->
<!--      + \color{red}{\beta_5X_{5\text{ Ethnicity}}} -->
<!--      + \color{red}{\beta_6X_{6\text{ Birth Weight}}} -->
<!--      + \color{purple}{\beta_7X_{7\text{ BMI Category}}} -->
<!--      \\ -->
<!--      &+& \color{blue}{\beta_8X_{8\text{ Age}}} -->
<!--      + \color{blue}{\beta_9X_{9\text{ Education}}} -->
<!--      + \color{blue}{\beta_{10}X_{10\text{ BMI Category}}}  -->
<!--      + \color{blue}{\beta_{11}X_{11\text{ Weight Gain during Pregnancy}}} -->
<!-- \end{eqnarray} -->
<!-- $$ -->



# Target Features

Within each omics layer, we will use the general linear model to screen for omics features that are associated with disease outcomes for liver injury.

A summary table for each omics layer will illustrate the numbers features within each layer, the number of possible pairwise comparisons, the significant pairwise comparisons, and the comparisons that are significant after adjusting for multiple comparisons.

## Linear Models

### ALT Sex-Specific
$$
Y_{\text{ALT}_{Sex-Specific}} = \alpha + \beta_1X_\text{Omics Feature} 
     + \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
$$

### Any Liver Enzyme 90p
$$
Y_{\text{(ALT|AST|GGT)}_{p90}} = \alpha + \beta_1X_\text{Omics Feature} 
     + \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
$$

### ALT Continuous
$$
Y_{\text{ALT}_{continuous}} = \alpha + \beta_1X_\text{Omics Feature} 
     + \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
$$

### AST Continuous
$$
Y_{\text{AST}_{continuous}} = \alpha + \beta_1X_\text{Omics Feature} 
     + \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
$$

### GGT Continuous
$$
Y_{\text{GGT}_{continuous}} = \alpha + \beta_1X_\text{Omics Feature} 
     + \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
$$

```{r}
# general liner model of dependent variable on independent var, adjusting for covariates
glm_y_x_cov <- function(dep_var, indep_var, i_family){

  # formula for univariate linear regression
  lin_reg_formula <- as.formula(paste0(dep_var, " ~ ", indep_var, " + ", paste0(collapse = " + ", names_covars)))
  # mylogit <- glm(admit ~ gre + gpa + rank, data = mydata, family = "binomial")
  
  multivar_reg <- df_selection_messy %>% 
    # format dichotomous outcome for sex-specific liver injury to 0,1 scale
    dplyr::mutate(i_alt_sex = as.double(i_alt_sex) - 1) %>% 
    glm(formula = lin_reg_formula, family = i_family)
  multivar_reg_sum <- summary(multivar_reg)
  p_val_beta <- multivar_reg_sum$coefficients[2,4]
  
  # single row of interesting data for purrr
  tibble(dep_var, indep_var, p_val_beta)

}
# glm_y_x_cov("i_alt_sex", "rs1976403_C", "binomial")
# glm_y_x_cov("pro_Adiponectin", "rs1976403_C", "gaussian")

# get all variables in class
get_names_byclass <- function(class_prefix){colnames(select(df_selection_messy, starts_with(class_prefix)))}
# get_names_byclass("pro_")

# get p-value for linear regression adjusting for covariates of a given outcome variable on all exposure variables in a class
get_lmyxcov_byclass <- function(class_prefix, dep_var, i_family){
  purrr::map_df(
    .x = get_names_byclass(class_prefix), 
    .f = glm_y_x_cov, dep_var = dep_var, i_family = i_family) %>% 
    dplyr::mutate(
      # adjust p-value for multiple comparisons
      p_val_fdr = p.adjust(p_val_beta, method = "BH"),
      i = row_number()
    ) %>% 
    # reorder variables
    dplyr::select(i, everything()) %>% 
    # filter by descending value of FDR p-value
    dplyr::arrange(p_val_fdr)
}
# get_lmyxcov_byclass("pro_", "i_alt_sex", "binomial") %>% summary
# get_lmyxcov_byclass("pro_", "rs1976403_C", "gaussian") %>% summary

# get p-value for general regression adjusting for covariates of combinations of all outcome variables in a class on all exposure variables in a second class
get_lmyxcov_by2class <- function(dep_class_prefix, indep_class_prefix, i_family){
  purrr::map_df(
    .x = get_names_byclass(dep_class_prefix), 
    .f = ~ purrr::map_df(.x = get_names_byclass(indep_class_prefix), .f = glm_y_x_cov, dep_var = .x, i_family = i_family)
    ) %>% 
    dplyr::mutate(
      # adjust p-value for multiple comparisons
      p_val_fdr = p.adjust(p_val_beta, method = "BH"),
      i = row_number()
    ) %>% 
    # reorder variables
    dplyr::select(i, everything()) %>% 
    # filter by descending value of FDR p-value
    dplyr::arrange(p_val_fdr)
}
# target_altsex_pro <- get_lmyxcov_by2class(dep_class_prefix = "i_alt_sex", indep_class_prefix = "pro_", i_family = "binomial")
# target_snp_pro <- get_lmyxcov_by2class(dep_class_prefix = "rs", indep_class_prefix = "pro_", i_family = "gaussian")
```


```{r}
# vector of liver injury outcome variables
liv_inj_var <- c("i_alt_sex", "i_q90_any", "alt_n", "ast_n", "ggt_n")
names(liv_inj_var)  <- c("ALT Sex-Specific", "Any Liver Enzyme 90p", "ALT Continuous", "AST Continuous", "GGT Continuous")

# identify target features that by
#   (1) all regression associations of a liver injury outcome on a single omics layer
#   (2) significant associations 
#   (3) significant associations after FDR adjustment
list_target_features <- function(dep_var, class_prefix, i_family){

  # regress ALT on all 36 inflammatory proteins
  all_reg <- get_lmyxcov_byclass(dep_var = dep_var, class_prefix = class_prefix, i_family = i_family)
  
  # all target features
  unique_indep_var <- tibble(unique(all_reg$indep_var))
  
  # features with significant associations BEFORE adjusting for multiple comparisons
  all_reg_sig <- all_reg %>% filter(p_val_beta < alpha_pval_beta)
  
  # features with significant associations AFTER adjusting for multiple comparisons
  all_reg_sig_fdr <- all_reg %>% filter(p_val_fdr < alpha_pval_beta)
  
  list(
    `Omics Features` = unique_indep_var,
    `Pairwise Comparisons (PWC)` = all_reg,
    `Significant PWC` = all_reg_sig,
    `Significant PWC, FDR Adjusted` = all_reg_sig_fdr
  )
  
}
# list_target_features("i_alt_sex", "pro", "binomial")


# get target features for all outcomes in a given omics layer
list_target_omics <- function(class_prefix){
  purrr::map2(
    .x = liv_inj_var,
    .y = c(rep("binomial", 2), rep("gaussian", 3)),
    .f = list_target_features,
    class_prefix = class_prefix
    )
}
# list_target_omics("pro")


# show summary table of numbers by outcome
summ_n_target <- function(target_list){
  bind_rows(
    lapply(target_list[[1]], nrow),
    lapply(target_list[[2]], nrow)
  )
  test <- map_dfr(.x = c(1:5), .f = ~ lapply(target_list[[.x]], nrow))
  bind_cols(Outcome = names(liv_inj_var), test)
}
# summ_n_target(list_target_omics("pro"))


# subset just the significant features from the target omics list
get_df_sig_feat <- function(target_list){
  bind_rows(
    # get the FDR-adjusted, significant associations,
    #    which is the fourth list element for every outcome list
    sapply(target_list, `[`, 4)
    ) %>% 
    dplyr::arrange(indep_var, dep_var)
}
# get_df_sig_feat(list_target_omics("pro"))


# show summary table of target linear regression for a given omics layer
html_scroll_sig_feat <- function(target_sig_feat_list){
  target_sig_feat_list %>% 
    kable() %>% kable_paper() %>% 
    scroll_box(width = "100%", height = "300px")
}
# html_scroll_sig_feat(get_df_sig_feat(list_target_omics("pro")))
```


## Proteomics

### Summary

```{r}
target_pro <- list_target_omics("pro")
summ_n_target(target_pro) %>% 
  kable() %>% kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
```

### Target Features
```{r}
target_pro_sig_feat <- get_df_sig_feat(target_pro)
target_pro_sig_feat %>% html_scroll_sig_feat()
```




## Serum Metabolomics

### Summary

```{r}
target_metser <- list_target_omics("metser")
summ_n_target(target_metser) %>% 
  kable() %>% kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
```

### Target Features
```{r}
target_metser_sig_feat <- get_df_sig_feat(target_metser)
target_metser_sig_feat %>% html_scroll_sig_feat()
```

## Urine Metabolomics

### Summary

```{r}
target_meturi <- list_target_omics("meturi")
summ_n_target(target_meturi) %>% 
  kable() %>% kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
```

### Target Features
```{r}
target_meturi_sig_feat <- get_df_sig_feat(target_meturi)
target_meturi_sig_feat %>% html_scroll_sig_feat()
```

# Summary of Targeted Omics Features

```{r}
# dataframe of all omics features that were significantly associated with omics features after adjusting for FDR
target_sig_fdr <- bind_rows(
  target_pro_sig_feat,
  target_metser_sig_feat,
  target_meturi_sig_feat
) %>% 
  # add factors for omics layers
  group_feature_prefix(indep_var) %>% 
  dplyr::mutate(
    # transform p-value to Manhattan like plot
    log_p_val_fdr = -log(p_val_fdr),
    # save a copy of long name for further coding
    full_name_indep_var = indep_var,
    # remove extraneous text from feature names
    indep_var = clean_feat_labs(indep_var),
    # clean liver injury names
    dep_var = factor(dep_var, levels = liv_inj_var, labels = names(liv_inj_var), ordered = T)
  )
```


## Heat Map

Note, any associations greater than `r alpha_pval_beta` significance are automatically excluded and show as white.

```{r}
# prepare data for heatmap
target_livinj_sig_fdr_heatmap_messy <- target_sig_fdr %>%
  # vars useful for heatmap
  dplyr::select(dep_var, indep_var, omics_group, p_val_fdr) %>%
  # transform p-values to -log(p-value, FDR-adjusted)
  mutate(p_val_fdr = -log(p_val_fdr)) %>% 
  # widen data so every SNP has a p-value (or missing value) for every omics feature
  pivot_wider(names_from = dep_var, values_from = p_val_fdr) %>% 
  # had to replace NA with zero, otherwise heatmap won't run
  replace(is.na(.), 0)

# heatmap requires matrix format, with no excess variables
target_livinj_sig_fdr_heatmap <- target_livinj_sig_fdr_heatmap_messy %>% 
  select(-omics_group, -indep_var) %>% # eliminate meta-info
  as.matrix()

# heatmap only accepts rownames, not column of rownames
rownames(target_livinj_sig_fdr_heatmap) <- target_livinj_sig_fdr_heatmap_messy$indep_var

# exponential color palette so zero values (previously NA) are white, with "hotter" indicating larger -log(p-val, FDR), i.e. smaller p-values, i.e. more significant associations
logpval_hotter_palette <- colorRampPalette(c("white", "orange", "red")) (n = 100)

gplots::heatmap.2(srtCol = 15,
  target_livinj_sig_fdr_heatmap, col = logpval_hotter_palette, key = TRUE, trace = "none", scale = "none",
  # add spectrum indicating which omics class to which each feature belongs
  RowSideColors = colors_fct_omics_dark[as.character(target_livinj_sig_fdr_heatmap_messy$omics_group)],
  dendrogram = "both", hclustfun = function(x) hclust(x, method = "ward.D")
)
```

## Manhattan Style Plot{.tabset}

```{r}
# function plot log p-value by omics features, faceted by each SNP
manhattan_facet <- function(df, xaxis_txt_size = 10){
  df %>% 
    ggplot(aes(x = indep_var, y = log_p_val_fdr, fill = omics_group)) +
    geom_col() +
    facet_wrap(facet = "dep_var") +
    scale_fill_manual(name = "", values = colors_fct_omics_dark) +
    geom_hline(yintercept = -log(alpha_pval_beta), linetype = 2, color = "red") +
    theme(
      legend.position = "bottom",
      axis.text.x = element_text(angle = 90, size = xaxis_txt_size)
      ) +
    ylab("-log(P-value FDR Adjusted)") +
    xlab("Omics Features")
}

# list of all SNPs with significant associations
# target_snp_sig_fdr$dep_var %>% unique
```

### All Outcomes of Liver Injury
```{r}
target_sig_fdr %>% manhattan_facet(xaxis_txt_size = 2)
```

### ALT Sex-Specific
```{r}
target_sig_fdr %>% filter(dep_var == "ALT Sex-Specific") %>% manhattan_facet()
```

### Any Liver Enzyme 90p
```{r}
target_sig_fdr %>% filter(dep_var == "Any Liver Enzyme 90p") %>% manhattan_facet()
```

### ALT Continuous
```{r}
target_sig_fdr %>% filter(dep_var == "ALT Continuous") %>% manhattan_facet()
```

### AST Continuous
```{r}
target_sig_fdr %>% filter(dep_var == "AST Continuous") %>% manhattan_facet()
```

### GGT Continuous
```{r}
target_sig_fdr %>% filter(dep_var == "GGT Continuous") %>% manhattan_facet()
```


# Partial Correlations

Partial correlations are by definition adjusted for at least one covariate. This procedure was produced after studying partial correlation [@kleinbaum_2013_applied] and metabolomics analysis [@alfano_2020_multiomic]. Pearson correlations were used.

1) Pearson's **Partial** Correlation Coefficient ($\rho_{XY|C_1, C_2,...,C_q}$) adjusting for covariates

2) F-test of $H_0:\rho_{XY|C_1, C_2,...,C_q} = 0$. But what hypothesis test does **ppcor** package use?

3) Multiple comparisons correction: False Discovery Rate

Unsure how to represent a partial correlation model mathematically, but it is based on the previous linear regression models.

$$
Y_\text{Target Feature} = \alpha + \beta_1X_\text{Omics Feature} 
     + \color{red}{\beta_iX_{i\text{ Child Sociodemographics}}}  
$$

```{r}
# get a vector of the targeted features within each omics layer (e.g. proteins, serum metabs, etc.)
get_target_features <- function(i_omics, liv_inj_outcome = names(liv_inj_var), df = target_sig_fdr){
  tar_feat_vec <- df %>% 
    dplyr::filter(dep_var %in% liv_inj_outcome) %>% 
    dplyr::filter(omics_group == i_omics) %>% 
    dplyr::select(full_name_indep_var) %>% 
    unique() %>% unlist()
  names(tar_feat_vec) <- NULL
  tar_feat_vec
}
# get_target_features("Serum Metabolites")
# get_target_features("Serum Metabolites", "AST Continuous")

# partial correlation of two variables, adjusting for covariates
pc_y_x_cov <- function(corr_var_1, corr_var_2){
  
  # if correlation variables are the same, return row of NA
  if(corr_var_1 == corr_var_2){
    
    # dataframe of 1 row with names of correlated variables and same number of NA as ppcor::pcor.test() output
    bind_cols(var_1 = corr_var_1, var_2 = corr_var_2)

   # run partial correlation if variables are different
  }else{
    
    # data for correlations
    data_messy <- df_selection_messy %>% 
      # select two variables for correlation and all covariates
      select(all_of(corr_var_1), all_of(corr_var_2), all_of(names_covars)) %>% 
      # covariates must be in numeric form for partial correlation function
      mutate(across(all_of(names_covars), ~as.numeric(.)))
  # return(data_messy) 
    # remove missing values, partial correlation functions do not accept NA
    data_clean <- data_messy %>% na.omit() 
  
    # Pearson partial correlation, adjusting for covariates
    pcor_out <- ppcor::pcor.test(method = "pearson",
      x = data_clean[,corr_var_1], 
      y = data_clean[,corr_var_2],
      z = data_clean[, names_covars]
      ) 
    
    # dataframe of 1 row with names of correlated variables and partial correlation output
    bind_cols(var_1 = corr_var_1, var_2 = corr_var_2, pcor_out)
  
  }

}
# pc_y_x_cov("pro_EGF", "pro_EGF") # case where attempt correlations on the same variable
# pc_y_x_cov("metser_log.PC.aa.C38.4", "pro_IL1beta") # case where correlation variables differ
# purrr::map_df(.x = get_target_features("Proteins"), .f = ~ purrr::map_df(.x = .x, .f = pc_y_x_cov, corr_var_2 = "pro_EGF"))

# Data frame of all Pearson correlations of:
#   1) Target features of a single, specified omics layer, and
#   2) All features from all omics layers
pcor_tarfeat_allfeat <- function(i_omics, liv_inj_outcome = names(liv_inj_var)){
  
  # get list of target features
  tar_feat <- get_target_features(i_omics, liv_inj_outcome = liv_inj_outcome)

  # Only continue with the output if there are target features
  if(length(tar_feat) > 0){
    
    # get list of all omics features
    all_feat <- names(dplyr::select(df_selection_messy, starts_with(c("pro_", "metser_", "meturi_"))))
    
    # dataframe of partial correlation output for each combination of target features with all features
    purrr::map_df(.x = tar_feat, .f = ~purrr::map_df(.x = all_feat, .f = pc_y_x_cov, corr_var_2 = .x)) %>%
      # identify target vars and all vars
      dplyr::rename(tar_feat = var_2, all_feat = var_1) %>% 
      # adjust p-value for multiple comparisons
      dplyr::mutate(p_val_fdr = p.adjust(p.value, method = "BH")) %>% 
      dplyr::arrange(p.value) %>% 
      # remove any comparisons that are not significant
      dplyr::filter(p_val_fdr < alpha_pval_beta)
    
  }else{
    message("There are no target features for this combination of liver injury outcome and multi-omics layer")
  }
  
}
# pcor_tarfeat_allfeat("Urine Metabolites")
# pcor_tarfeat_allfeat("Urine Metabolites", liv_inj_outcome = "ALT Continuous")
# pcor_tarfeat_allfeat("Urine Metabolites", liv_inj_outcome = "ALT Continuous")
```

```{r}
# # list of all correlations of targeted features
# pcor_tar_feat <- list(
#   pcor_tar_pro = pcor_tarfeat_allfeat("Proteins"),
#   pcor_tar_metser = pcor_tarfeat_allfeat("Serum Metabolites"),
#   pcor_tar_meturi = pcor_tarfeat_allfeat("Urine Metabolites")
# )
# save(pcor_tar_feat, file = paste0(dir_data_secure, "/partial_corr_target_features_", Sys.Date(), ".Rda"))

# takes a while to run, so save it 
load(paste0(dir_data_secure, "/partial_corr_target_features_2021-02-03.Rda"))

# filter partial correlations by target features belonging to a specific liver injury outcome
pcor_by_liv_inj <- function(liv_inj_outcome){
    map2(
      .x = pcor_tar_feat, # list of pcor for each omics layer with all target features combined
      .y = c("Proteins", "Serum Metabolites", "Urine Metabolites"),
      .f = ~ dplyr::filter(
        .x , 
        tar_feat %in% get_target_features(.y, liv_inj_outcome = liv_inj_outcome))
      )
}

pcor_tar_feat_altsexs <- pcor_by_liv_inj("ALT Sex-Specific")
pcor_tar_feat_anylivi <- pcor_by_liv_inj("Any Liver Enzyme 90p")
pcor_tar_feat_contalt <- pcor_by_liv_inj("ALT Continuous")
pcor_tar_feat_contast <- pcor_by_liv_inj("AST Continuous")
pcor_tar_feat_contggt <- pcor_by_liv_inj("GGT Continuous")
```



# Circos Globe Plots

We implemented Circos plots to visualize partial correlations for each targeted analysis [@gu_2014_circlize]. 

Targeted features tended to correlated with features in the same omics layer, which cluttered the correlations between target features and features in the other omics layers. Correlations between features belonging to the same omics layer were removed to clarify the more interesting associations.

The colors of the links represent the value of the partial correlations where red is positive, blue is negative, and colors fade to white as they approach 0. Most correlations were between ±0.6, which is where the maximum colors were set to visually appreciate the differences in the smaller correlations. The legend shows the extreme values of correlations as well as the theoretical limits of ±1; colors are also shown for medians of positive and negative correlations.

```{r}
# create circos plot for each omics layer
p_circos_omics <- function(pcor_list = pcor_tar_feat, omics_layer){
  
  # if there's no target features, exit the function early with message
  if(nrow(pcor_list[[paste0("pcor_tar_", omics_layer)]]) == 0){
    
    message("There are no target features for this combination of liver injury outcome and multi-omics layer")
    
    }else{
  
      # omics layers in order
      omics_layer_prefix <- c("pro", "metser", "meturi")
      omics_layer_order <- c("Proteins", "Serum Metabolites", "Urine Metabolites")
      metser_class_order <- sort(unique(lst_metabolome_serum$names_metabolome_serum$mtb_metabolites_serum$class))
      group_order <- c(omics_layer_order[1], metser_class_order, omics_layer_order[3])
      
      # serum metabolite classes for each feature
      metser_class <- lst_metabolome_serum$names_metabolome_serum$mtb_metabolites_serum$class
      names(metser_class) <- clean_punct(clean_feat_aa_ae(clean_feat_labs(lst_metabolome_serum$names_metabolome_serum$mtb_metabolites_serum$feat_label)))
      
      # select parameters for single omics layer used in function
      i_params <- tibble(prefix = omics_layer_prefix, string = omics_layer_order) %>% dplyr::filter(prefix == omics_layer)
    
      # names of target features
      names_tar_feat <- clean_punct(clean_feat_aa_ae(clean_feat_labs(unique(pcor_list[[paste0("pcor_tar_", i_params$prefix)]]$tar_feat))))
      
      # single dataframe for circlize plot
      df_messy <- pcor_list[[paste0("pcor_tar_", i_params$prefix)]] %>% 
        # add metabolite classes to target dataframe
        left_join(by = "all_feat", 
                  y = dplyr::rename(select(lst_metabolome_serum$names_metabolome_serum$mtb_metabolites_serum,
                                           feat_label, class, common_name), 
                                    all_feat = feat_label)) %>% 
        # add grouping variable identifying omics layer
        group_feature_prefix(all_feat) %>% 
        # remove correlations within the target omics layer (too distracting)
        dplyr::filter(omics_group != i_params$string) %>%
        # add identity rows for target features
        bind_rows(data.frame(
          tar_feat = names_tar_feat, 
          all_feat = names_tar_feat, 
          estimate = rep(0, length(names_tar_feat)), 
          omics_group = rep(i_params$string, length(names_tar_feat)),
          # assign class if serum metabolite analysis, otherwise leave blank
          class = ifelse(omics_layer == "metser", 
                         metser_class[names_tar_feat], "")
          ))
    
      # format data for circlize
      df_details <- df_messy %>% 
        dplyr::mutate(
          # format omics grouping in consistent order, drop unused levels
          omics_group_class = ifelse(omics_group == "Serum Metabolites", 
                                     class, omics_group) %>% 
            factor(levels = group_order, ordered = TRUE) %>% 
            droplevels(),
          # remove long prefixes from names
          tar_feat = clean_punct(clean_feat_aa_ae(clean_feat_labs(tar_feat))),
          all_feat = clean_punct(clean_feat_aa_ae(clean_feat_labs(all_feat))),
          # indicator for target features, helps with ordering
          target_feat = ifelse(all_feat %in% names_tar_feat, 1, 2) %>% factor(levels = 1:2, labels = c("Target", "Correlation"))
          ) %>% 
        # arrange target features first, then magnitudes of association
        dplyr::arrange(omics_group_class, target_feat, estimate)
    
      # select variables for chord diagram
      df <- df_details %>% select(tar_feat, all_feat, estimate, omics_group, omics_group_class)
    
        # vector of features by group, formatted for circlize
        vec_group <- df$omics_group_class
        names(vec_group) <- df$all_feat
     
        # dataframe of feature colors
        df_grid_col <- df %>% select(all_feat, omics_group, omics_group_class) %>% 
          mutate(grid.col = case_when(
            # replace target features with dark colors
            all_feat %in% clean_feat_aa_ae(clean_feat_labs(names_tar_feat)) ~ colors_fct_omics_dark[i_params$string],
            # add serum metabolites light green color if serum metabolite and class does not match broad omics groups
            omics_group == "Serum Metabolites" & !(omics_group_class %in% unique(.data$omics_group)) ~ colors_fct_omics_light[["Serum Metabolites"]],
            # add light colors for all omics features
            TRUE ~ colors_fct_omics_light[as.character(omics_group_class)]
            ))
        # vector of colors for grid blocks, formatted for circlize
        grid.col <- df_grid_col$grid.col
        names(grid.col) <- df_grid_col$all_feat
       
        # color pal indicating strength of correlation
        #   always set extreme values at ±1 for blue-white-red colors
        col_pal_estimate <- colorRamp2(c(-0.6, 0, 0.6), c("blue", "white", "red"))
    
        # vector of link colors, formatted for circlize
        link.col <- col_pal_estimate(df$estimate)
        names(link.col) <- df$all_feat
        
        # data frame of group colors
        df_group <- unique(df %>% select(omics_group, omics_group_class)) %>% 
          dplyr::mutate(color = colors_fct_omics_light[.data$omics_group])
       
        # initialize chord diagram
        par(xpd = NA)
        circos.par(start.degree = 90)
        chordDiagram(df, annotationTrack = c("grid"), group = vec_group, grid.col = grid.col, col = link.col, preAllocateTracks = 1, )
      
        # text for serum metabolomics subgroups
        group_names <- function(n){
          # text for serum groups
          group_vec <- vec_group[vec_group == df_group$omics_group_class[n]]
          highlight.sector(names(group_vec), track.index = 1, text = str_to_title(as.character(group_vec[n])), col = df_group$color[n],
                           text.col = "black", facing = "bending", niceFacing = TRUE, text.vjust = "0mm", cex = 0.5, 
                           padding = c(0.075, 0, -0.9, 0))
        }
        lapply(1:length(unique(vec_group)), group_names)
      
        # https://stackoverflow.com/questions/31943102/rotate-labels-in-a-chorddiagram-r-circlize
        circos.trackPlotRegion(track.index = 1, panel.fun = function(x, y) {
          
          circos.text(CELL_META$xcenter, CELL_META$ylim[1] + 0.15, CELL_META$sector.index, cex = 0.5,
                      facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
          }, bg.border = NA)
        
        # legend omics layers
        legend(title = "Omics Layer", bty = "n", x = "topright", cex = 0.8, pch = 15, pt.cex = 1,
               col = c(df_group$color), 
               legend = c(str_to_title(as.character(df_group$omics_group_class)))
               )
      
        # legend links exposures
        # legend is responsive, so shows colors for maximum values
        est_max <- max(df$estimate)
        est_med_u <- median(dplyr::filter(df, estimate >= 0)$estimate)
        est_med_l <- median(dplyr::filter(df, estimate < 0)$estimate)
        est_min <- min(df$estimate)
        range_scale <- c(1, est_max, est_med_u, 0, est_med_l, est_min, -1)
        legend(title = "Correlation", bty = "n", x = "bottomright", cex = 0.8, pch = 15, pt.cex = 1, xjust = 1,
               col = col_pal_estimate(range_scale),
               legend = paste0(c(rep(" ", 4), rep("", 3)),
                 paste(sep = "   ",
                       formatC(range_scale, digits = 2, format = "f"), 
                       c("", "Maximum", "Median Positive", "None", "Median Negative", "Minimum", ""))
                 )
               )
        
        # info details
        # circos.info()
        
        # close commands of the circos plot
        circos.clear()
        
    }
  
}
# p_circos_omics(pcor_list = pcor_tar_feat_anylivi, omics_layer = "pro")
# p_circos_omics(pcor_list = pcor_tar_feat_contast, omics_layer = "meturi")
```

## Proteomics{.tabset}

### ALT Sex-Specific

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_altsexs_pro.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_altsexs, omics_layer = "pro")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_altsexs_pro.png)

### Any Liver Enzyme 90p

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_anylivi_pro.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_anylivi, omics_layer = "pro")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_anylivi_pro.png)

### ALT Continuous

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_contalt_pro.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contalt, omics_layer = "pro")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contalt_pro.png)

### AST Continuous

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_contast_pro.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contast, omics_layer = "pro")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contast_pro.png)

### GGT Continuous

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_contggt_pro.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contggt, omics_layer = "pro")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contggt_pro.png)

### All Outcomes of Liver Injury

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_pro.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(omics_layer = "pro")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_pro.png)


## Serum Metabolomics{.tabset}

### ALT Sex-Specific

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_altsexs_metser.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_altsexs, omics_layer = "metser")
dev.off() %>% invisible()
```
<!-- ![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_altsexs_metser.png) -->
There are no target features for this combination of liver injury outcome and multi-omics layer

### Any Liver Enzyme 90p

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_anylivi_metser.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_anylivi, omics_layer = "metser")
dev.off() %>% invisible()
```
<!-- ![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_anylivi_metser.png) -->
There are no target features for this combination of liver injury outcome and multi-omics layer

### ALT Continuous

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_contalt_metser.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contalt, omics_layer = "metser")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contalt_metser.png)

### AST Continuous

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_contast_metser.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contast, omics_layer = "metser")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contast_metser.png)

### GGT Continuous

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_contggt_metser.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contggt, omics_layer = "metser")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contggt_metser.png)

### All Outcomes of Liver Injury

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_metser.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(omics_layer = "metser")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_metser.png)


## Urine Metabolomics{.tabset}

### ALT Sex-Specific

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_altsexs_meturi.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_altsexs, omics_layer = "meturi")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_altsexs_meturi.png)


### Any Liver Enzyme 90p
```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_anylivi_meturi.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_anylivi, omics_layer = "meturi")
dev.off() %>% invisible()
```
<!-- ![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_anylivi_meturi.png) -->
There are no target features for this combination of liver injury outcome and multi-omics layer

### ALT Continuous

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_contalt_meturi.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contalt, omics_layer = "meturi")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contalt_meturi.png)

### AST Continuous

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_contast_meturi.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contast, omics_layer = "meturi")
dev.off() %>% invisible()
```
<!-- ![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contast_meturi.png) -->
There are no target features for this combination of liver injury outcome and multi-omics layer

### GGT Continuous

```{r, message=FALSE}
png(file=paste0(dir_report, "/p_circos_target_contggt_meturi.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(pcor_list = pcor_tar_feat_contggt, omics_layer = "meturi")
dev.off() %>% invisible()
```
<!-- ![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_contggt_meturi.png) -->
There are no target features for this combination of liver injury outcome and multi-omics layer

### All Outcomes of Liver Injury

```{r, message=FALSE}
# save high resolution image
png(file=paste0(dir_report, "/p_circos_target_meturi.png"), res = 350, width = 11, height = 8.5, units = "in")
p_circos_omics(omics_layer = "meturi")
dev.off() %>% invisible()
```
![](/Volumes/GoogleDrive/My Drive/20200601_Postdoc_Envn_Omics_NAFLD/HELIX NAFLD/3 Reports/p_circos_target_meturi.png)



# Enriched Omics Features

# mQTL Analysis

Upload SNP rsID to [mQTL Database](mtqldb.org) with the following parameters:  
* Database: MatrixEQTL 
* Timepoint: Childhood
* Trans. Distance (bp) 1 000 000 = 1 000 Mb

Download CpG sites as .CSV

Import CpG sites into R

```{r}

```



# References
