---
title: "Exploratory Data Analysis"
author: "Dom Grisafe"
date: "`r Sys.Date()`"
output:
  html_document:
    keep_md: FALSE
    toc: true
    toc_float: true
    toc_depth: 2
csl: hepatology.csl
bibliography: "/Users/domusc/Dropbox/Zotero_Library/!ZoteroLibrary.bib"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(styledom)
library(tidyverse)
library(cowplot)
  theme_set(theme_cowplot())
library(labelled)

# run this if knitr throws terminal error:
# *** caught segfault *** address 0x7ff4d479fba8, cause 'memory not mapped'
installed.packages("tableone", "kableExtra")
  library(tableone)
  library(kableExtra)
  
# SNP data packages
library(GenomicRanges)
library(BiocGenerics)
library(GenomeInfoDb)
library(stats4)
library(BioCircos) # visualize SNPs in Circos plots
  
source("!load_clean_data.R")
```

# Color Key{.tabset}

```{r}
show_colors <- function(color_pal, label_name){
  palette(color_pal)
  barplot(rep(1, length(color_pal)), col = 1:length(color_pal), names.arg = str_wrap(names(color_pal), 9), main = label_name, horiz = F, axes = F, las = 0)
}
```

## Liver Injury
```{r}
show_colors(color_pal = colors_fct_liv_inj, 
            label_name = "Liver Injury"
            # label_name = var_label(liv_enz_data$liver_enz$i_q90_any)
            )
```

## HELIX Cohorts
```{r}
show_colors(color_pal = colors_fct_cohorts, label_name = "Cohorts")
```

## Omics Features

```{r}
show_colors(color_pal = colors_fct_omics_dark, label_name = "Omics Biomarker Features (Dark)")
```

```{r}
show_colors(color_pal = colors_fct_omics_light, label_name = "Omics Biomarker Features (Light)")
```



# Outcome: Liver Injury

Liver injury was defined as:

1) **Primary Outcome** — NAFLD liver injury defined as ALT ≥ 22.1 U/L for girls and ≥ 25.8 U/L for boys [@schwimmer_2010_safety].  

2) **Secondary Outcome** — Any liver enzyme (ALT, AST, or GGT) ≥ 90th percentile within HELIX cohort, which has been used previously in HELIX [@stratakis_2020_prenatal].  

```{r}
# create dataframe with the 90th percentile cutoffs for each liver injury variables
q90_liv_enz <- liv_enz_data$liver_enz %>% 
  dplyr::summarise(
    q90_alt_n  = quantile(alt_n, 0.9, na.rm = TRUE),
    q90_ast_n  = quantile(ast_n, 0.9, na.rm = TRUE),
    q90_ggt_n  = quantile(ggt_n, 0.9, na.rm = TRUE),
    q90_ck18_n = quantile(ck18_n, 0.9, na.rm = TRUE)
  )

# create histograms of liver enzymes illlustrating 90th percentile
hist_liv_enz <- function(liv_enz, df = liv_enz_data$liver_enz){
  
  # formatting dplyr variables
  liv_enz_name <- enquo(liv_enz) 
  
  # name of 90th quartile cutoff
  q90_liv_enz_name <- paste0("q90_", rlang::quo_get_expr(liv_enz_name))
  
  # histogram with SD bin widths and vertical line of enzyme 90th percentile
  df %>%
    dplyr::mutate(i_fill = ifelse(as.double(!!liv_enz_name) >= as.double(q90_liv_enz[q90_liv_enz_name]), "Injury", "Normal")) %>% 
    ggplot(aes(x = !!liv_enz_name)) +
    geom_histogram(aes(fill = i_fill), bins = 20, na.rm = TRUE) +
    geom_vline(xintercept = as.double(q90_liv_enz[q90_liv_enz_name]), linetype = 2, color = "red") +
    ggtitle(toupper(str_remove(rlang::quo_get_expr(liv_enz_name), "_n"))) +
    scale_y_continuous(limits = c(0, 650)) +
    scale_fill_manual(name = "Liver", values = rev(colors_fct_liv_inj), guide = guide_legend(reverse = TRUE)) +
    theme(legend.position = "none")
}
```

## Histograms{.tabset}


### ALT Sex-Specific

```{r}
# histogram with SD bin widths and vertical line of enzyme 90th percentile
p_alt_sex <- df_selection_messy %>% 
  ggplot(aes(x = alt_n)) +
  geom_histogram(aes(fill = i_alt_sex), bins = 20, na.rm = TRUE) +
  scale_y_continuous(limits = c(0, 250)) +
  scale_fill_manual(name = "Liver", values = rev(colors_fct_liv_inj), guide = guide_legend(reverse = TRUE)) +
  theme(legend.position = "bottom")
p_alt_sex
```

```{r}
p_alt_sex +
  facet_wrap(~e3_sex_None) +
  geom_vline(
    data = filter(df_selection_messy, e3_sex_None == "female"), 
    aes(xintercept = 22), linetype = 2, color = "red"
    ) +
  geom_vline(
    data = filter(df_selection_messy, e3_sex_None == "male"), 
    aes(xintercept = 25), linetype = 2, color = "red"
    )
```


### Enzyme ≥ 90th Percentile

```{r}
ggpubr::ggarrange(
  hist_liv_enz(alt_n),
  hist_liv_enz(ast_n),
  hist_liv_enz(ggt_n),
  hist_liv_enz(ck18_n)
)
```


## Barplots{.tabset}

```{r}
# reshape data into long format for creating bar graphs
liver_enz_p_data <- liv_enz_data$liver_enz %>% 
  pivot_longer(cols = starts_with("i_"), names_prefix = "^i_", names_to = c("enzyme")) %>% 
  dplyr::rename(Liver = value) %>% 
  dplyr::mutate(enzyme = fmt_liv_enz_ord(enzyme))

# tibble illustrating the differences in numbers from liver injury definitions
liver_enz_p_data %>% 
  group_by(Liver, enzyme) %>% 
  summarise(count = n(), .groups = "keep") %>% 
  ungroup() %>% 
  pivot_wider(id_cols = enzyme, names_from = Liver, values_from = count) %>%
  dplyr::arrange(enzyme) %>% styledom::quick_kable()
```

### Overall

```{r}
# calculate percentages
txtpct_liver_enz_p_data <- liver_enz_p_data %>% 
  group_by(enzyme) %>% 
  mutate(n_cohort = n()) %>% 
  count(Liver, n_cohort) %>% 
  mutate(pct = n / n_cohort) 

# plot
plot_bar_liver_enz <- txtpct_liver_enz_p_data %>% dplyr::mutate(enzyme = fct_rev(enzyme)) %>% 
  ggplot(aes(x = enzyme, y = n, fill = Liver, label = scales::percent(pct))) +
  geom_col(position=position_stack(reverse = TRUE)) +
  xlab("Liver Enzymes") +
  ylab(paste0("Children     n = ", nrow(liv_enz_data$liver_enz))) +
  coord_flip() +
  scale_fill_manual(values = colors_fct_liv_inj) +
  theme(legend.position = "bottom") +
  geom_text(position = position_stack(reverse = TRUE, vjust = 0.5))

  
# show plot overall with features differing from plot stratified by cohort 
plot_bar_liver_enz +
  scale_y_continuous(limits = c(0, 1105), breaks = seq(0,1100, 250))
```


### Individuals

```{r}
# plot participants with liver injury defined by different liver enzymes
set.seed(2020)
liver_enz_p_data %>% 
  ggplot(aes(
    x = factor(helixid, ordered = T, arrange(liv_enz_data$liver_enz, i_alt_sex, i_q90_any, i_q90_alt, i_q90_ast, i_q90_ggt, i_q90_ck18)[["helixid"]]), 
    y = fct_rev(enzyme), 
    color = Liver
    )) +
  geom_jitter(alpha = 0.4) +
  scale_color_manual(values = colors_fct_liv_inj) +
  theme(
    axis.text.x = element_blank(), 
    axis.ticks.x = element_blank(),
    legend.position = "bottom"
    ) +
  ylab("Liver Enzymes") +
  xlab(paste0("Children     n = ", nrow(liv_enz_data$liver_enz)))
```


### By Cohort

```{r}
# summary tables for labels
txtpct_liver_enz_p_data_cht <- liver_enz_p_data %>% 
  group_by(h_cohort, enzyme) %>% 
  mutate(n_cohort = n()) %>% 
  count(h_cohort, Liver, n_cohort) %>% 
  mutate(pct = n / n_cohort) 

# plot
plot_bar_liver_enz_cht <- txtpct_liver_enz_p_data_cht %>% 
  ggplot(aes(x = fct_rev(enzyme), y = n, fill = Liver, label = scales::percent(pct, accuracy = 2))) +
  geom_col(position=position_stack(reverse = TRUE)) +
  facet_wrap(vars(h_cohort), ncol = 2) +
  xlab("Liver Enzymes") +
  ylab(paste0("Children     n = ", nrow(liv_enz_data$liver_enz))) +
  coord_flip() +
  scale_fill_manual(values = colors_fct_liv_inj) +
  theme(legend.position = "bottom") +
  geom_text(position = position_stack(reverse = TRUE, vjust = 0.5))

  
# show plot overall with features differing from plot stratified by cohort 
plot_bar_liver_enz_cht +
  scale_y_continuous(limits = c(0, 271), breaks = seq(0, 250, 50))
```


# Covariates: Sociodemographics

HELIX sociodemographic characteristics during **pregnancy** and **postnatal** periods. 

All covariables were identified from HELIX data dictionary (Domain == "Covariates").

## Table 1{.tabset}

```{r}
# function that prints table 1, stratified by different liver injury variables
get_table1_liv_inj <- function(df, liv_inj_strata, i_select = FALSE){
  table1_vars <- colnames(select(df, -helixid, -starts_with("decile_")))
  
  table1_html <- CreateTableOne(
    data = exposome_covs$table1_data, 
    strata = liv_inj_strata,
    vars = table1_vars,
    includeNA = TRUE
    ) %>% 
    print(exact = "stage", quote = FALSE, noSpaces = F, printToggle = F, varLabels = T) %>%
    kableone(digits = 3) %>% 
    kable_styling(bootstrap_options = c("striped", "hover"))
   if(i_select){
     table1_html %>% 
       pack_rows("Cohort", 2, 8) %>% 
       pack_rows("Outcome: Liver Enzymes in Child", 9, 17) %>% 
       pack_rows("Child Sociodemographics", 18, 31) %>% 
       pack_rows("Child Modifiable Risk Factors", 32, 41) %>% 
       pack_rows("Parental Sociodemographics", 42, 58)
       
   }else{
     table1_html %>% 
       pack_rows("Cohort", 2, 8) %>% 
       pack_rows("Outcome: Liver Enzymes in Child", 9, 17) %>% 
       pack_rows("Pregnancy Covariates", 18, 95) %>% 
       pack_rows("Postnatal Covariates", 96, 150)
     }
}
```


### Select Covariates{.tabset}

Discussed covariate selection with NS on 11/20/2020.

Focus on sociodemographics of child, mother, and father. 

Also include child BMI and diet covariates associated with liver injury. These covariates may be intermediate pathways for liver injury. Unsure whether to include these in final model, but carry them forward for future analyses. May perform sensitivity analysis to assess whether including these variables changes outcome.  


Discussed with lab group on 12/15/2020.

Likely over-adjusting. Adjust for cohort, sex, and age (minimal covariates). But will show variables discussed above in Table 1.


#### ALT Sex-Specific
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "i_alt_sex", TRUE)
```

#### Enzyme ≥ 90th Percentile{.tabset}

##### Any Enzyme
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "i_q90_any", TRUE)
```

##### ALT
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "i_q90_alt", TRUE)
```

##### AST
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "i_q90_ast", TRUE)
```

##### GGT
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "i_q90_ggt", TRUE)
```

##### CK-18
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "i_q90_ck18", TRUE)
```

#### Country Cohort
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data_select, liv_inj_strata = "h_cohort", TRUE)
```

#### Dictionary
```{r}
# create data dictionary with variable names and descriptions
data.frame(
  Variable_name_TRANS = colnames(exposome_covs$table1_data_select), 
  description = c(
    exposome_covs$data_dict_id_covs$description[exposome_covs$data_dict_id_covs$Variable_name_TRANS=="HelixID"],
    exposome_covs$data_dict_id_covs$description[exposome_covs$data_dict_id_covs$Variable_name_TRANS=="h_cohort"],
    unlist(var_label(exposome_covs$table1_data_select)))
  ) %>% 
  quick_kable() %>% 
  pack_rows("Identification", 1, 1) %>% 
  pack_rows("Cohort", 2, 2) %>% 
  pack_rows("Outcome: Liver Enzymes in Child", 3, 11) %>% 
  pack_rows("Child Sociodemographics", 12, 15) %>% 
  pack_rows("Child Modifiable Risk Factors", 16, 19) %>% 
  pack_rows("Parental Sociodemographics", 20, 26)
```


### All HELIX Covariates{.tabset}

#### ALT Sex-Specific
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "i_alt_sex")
```

#### Enzyme ≥ 90th Percentile{.tabset}

##### Any Enzyme
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "i_q90_any")
```

##### ALT
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "i_q90_alt")
```

##### AST
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "i_q90_ast")
```

##### GGT
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "i_q90_ggt")
```

##### CK-18
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "i_q90_ck18")
```

#### Country Cohort
```{r}
get_table1_liv_inj(df = exposome_covs$table1_data, liv_inj_strata = "h_cohort")
```

#### Dictionary

```{r}
# create data dictionary with variable names and descriptions
data.frame(
  Variable_name_TRANS = colnames(exposome_covs$table1_data), 
  description = c(
    exposome_covs$data_dict_id_covs$description[exposome_covs$data_dict_id_covs$Variable_name_TRANS=="HelixID"],
    exposome_covs$data_dict_id_covs$description[exposome_covs$data_dict_id_covs$Variable_name_TRANS=="h_cohort"],
    unlist(var_label(exposome_covs$table1_data)))
  ) %>% 
  quick_kable() %>% 
  pack_rows("Identification", 1, 1) %>% 
  pack_rows("Cohort", 2, 2) %>% 
  pack_rows("Outcome: Liver Enzymes in Child", 3, 12) %>% 
  pack_rows("Pregnancy Covariates", 13, 55) %>% 
  pack_rows("Postnatal Covariates", 56, 94)
```


## Collinearity{.tabset}

Are outcome and covariate variables colinear?

Correlation coefficients are colored with the following scale:  

* $\rho > 0.5$          <span style="color: red;">red</span>
* $0.3 > \rho <  0.5$   <span style="color: lightcoral;">light red</span>
* $-0.3 > \rho <  0.3$  black 
* $-0.5 > \rho < -0.3$  <span style="color: lightblue;">light blue</span>
* $\rho < -0.5$         <span style="color: blue;">blue</span>

```{r}
# function to create Spearman correlation table with colors indicating strength of association
corr.tab <- function(df){
  table <- round(cor(df %>% mutate_all(as.double), method = "spearman"), 2)  
  pretty_table <- tibble(
    variable = rownames(table),
    table %>% 
    as_tibble() %>%
    mutate(across(1:ncol(table), ~ kableExtra::cell_spec(sprintf("%.3f", round(., 3)), "html",
      color = ifelse(. >  0.5, "red",
              ifelse(. <  0.5 & . >  0.3, "lightcoral",
              ifelse(. <  0.3 & . > -0.3, "black",
              ifelse(. < -0.3 & . > -0.5, "lightblue",
              "blue"
                     ))))
      )))
    )
    
    # print html table
    pretty_table %>% quick_kable()
}
# corr.tab(covs_coh_liv)
```


### Cohort and Liver Enzymes in Child

Liver enzymes look colinear, but this should be accounted for by the categorized outcome variable.

```{r}
covs_coh_liv <- exposome_covs$table1_data_select %>% 
  select(h_cohort, i_alt_sex, i_q90_any, ends_with("_n")) 
covs_coh_liv %>% pairs
corr.tab(covs_coh_liv)
```

### Child Sociodemographics

French cohort (EDEN) looks older than the other groups.

```{r}
covs_chi_socdem <- exposome_covs$table1_data_select %>% 
  select(h_cohort, hs_child_age_days_None, e3_sex_None, h_ethnicity_c_None, e3_bwc_None)
covs_chi_socdem %>% pairs
corr.tab(covs_chi_socdem)
```

### Child Modifiable Risk Factors

```{r}
covs_chi_modify <- exposome_covs$table1_data_select %>% 
  select(h_cohort, hs_bmicat_None, hs_c_bmi_None, h_sugar_post_Log, h_nonalc_post_Log) 
covs_chi_modify %>% pairs
corr.tab(covs_chi_modify)
```

### Parental Sociodemographics

```{r}
covs_par_sociodem <- exposome_covs$table1_data_select %>% 
  select(h_cohort, h_age_None, h_fage_None, h_edumc_None, h_edufc_None, h_mbmi_None, h_mbmic_None, hs_wgtgain_None)
covs_par_sociodem %>% pairs
corr.tab(covs_par_sociodem)
```

# Exposure: Genetic SNPs{.tabset}

How were these SNPs selected? Was provided list of SNPs from previous researchers.

Visualize SNPs with [Circos plots](https://cran.r-project.org/web/packages/BioCircos/vignettes/BioCircos.html)

Single SNP was unable to be located in reference genome using : [rs657152](https://www.snpedia.com/index.php/Rs657152)


## Circos Plot

```{r}
# Chromosomes on which the points should be displayed
points_chromosomes <- as.character(seqnames(g_snp_genome$g_snps_ranges))

# Colors indicating whether SNP is NAFLD-specific or general to liver injury
points_colors <- ifelse(g_snp_genome$names_snp$g_snps_details$type_liv_inj == "NAFLD-specific", "tomato2", "darkblue")

# Coordinates on which the points should be displayed
points_coordinates <- as.numeric(pos(g_snp_genome$g_snps_ranges))

# Values associated with each point, used as radial coordinate 
#   on a scale going to minRadius for the lowest value to maxRadius for the highest value
points_values = rep(1:11, 11)

tracklist <- BioCircosSNPTrack('mySNPTrack', points_chromosomes, points_coordinates, 
  points_values, colors = points_colors, minRadius = 0.5, maxRadius = 0.9)

# Background are always placed below other tracks
tracklist <- tracklist + BioCircosBackgroundTrack("myBackgroundTrack", 
  minRadius = 0.5, maxRadius = 0.9,
  borderColors = "#AAAAAA", borderSize = 0.6, fillColors = "#B3E6FF")  

# get chromosome lengths and names from **library(rCGH)** in  **get_human_genome.R**
genome_hg38 <- rCGH::hg38$length
names(genome_hg38) <- rCGH::hg38$chrom %>% 
  str_replace("23", "X") %>% str_replace("24", "Y") # assign names to sex xsome
genome_hg38 <- as.list(genome_hg38) # format for plotting function

# plot Circos
BioCircos(tracklist = tracklist, 
  chrPad = 0.05, displayGenomeBorder = FALSE, yChr =  FALSE,
  genomeTicksDisplay = FALSE,  genomeLabelTextSize = 18, genomeLabelDy = 0,
  genome = genome_hg38, zoom = FALSE
  )

# legend info
cir_leg_col <-  unique(points_colors)
cir_leg_txt <-  levels(g_snp_genome$names_snp$g_snps_details$type_liv_inj)
```

**Legend**

Phenotypic Outcomes for SNPs:  

* <span style='color: tomato;'>NAFLD-specific</span>  
* <span style='color: darkblue;'>General Liver Injury</span>  


## Table

```{r}
snp_table <- g_snp_genome$names_snp$g_snps_details
snp_table_pretty <- snp_table %>% 
  dplyr::arrange(type_liv_inj, chr, snp) %>% 
  dplyr::mutate(
    j_snps = 1:nrow(snp_table), # create index column to identify n SNPs
    allele_risk_ref = paste(ref_allele, minor_allele, sep = "/")
    ) %>%
  dplyr::select(j_snps, gene, snp, chr, position, allele_risk_ref, reference) 
  
#indexes for grouping rows
lvl_1 <- levels(snp_table$type_liv_inj)[1]
n_lvl_1 <- length(snp_table$type_liv_inj[snp_table$type_liv_inj == lvl_1])
lvl_2 <- levels(snp_table$type_liv_inj)[2]
n_lvl_2 <- length(snp_table$type_liv_inj[snp_table$type_liv_inj == lvl_2])

# html table of SNP info
snp_table_pretty %>% 
  quick_kable() %>% 
  kableExtra::pack_rows(lvl_1, 1, n_lvl_1) %>% 
  kableExtra::pack_rows(lvl_2, 1 + n_lvl_1, n_lvl_2)
```



# Missing Data Analysis

Need to add visuals demonstrating missing data


# Univariate Analysis

Volcano plots are used to illustrate differential expression of omics features between the two groups: children with liver injury and healthy normal children.

```{r}
# create volcano plots, be able to change outcome variable
plot_volcano <- function(
  df = df_selection_messy, 
  prefix, i_q90_liv_injury, 
  mag_diff = 1.1, alpha = 0.05, trim_label = " "
  ){
  dplyr_i_q90_liv_injury <- enquo(i_q90_liv_injury)
  
  # cutoff limits for vertical cutoffs of significance
  log2_mag_diff <- log(mag_diff, base = 2)
  
  # dataframe of just proteomics data
  df_pro_volcano <- df %>%
    select(helixid, !!dplyr_i_q90_liv_injury, starts_with(prefix))
  
  # if normal, use parametric test, otherwise non-parametric test
    # Shapiro-Wilk test for normality (p < 0.5 suggests not normal)
    # df_pro_volcano %>% 
    #   dplyr::select(starts_with(prefix)) %>% 
    #   map_df(.f = ~ unlist(shapiro.test(.)))
  
  # create data for volcano plot
  pro_means_messy <- df_pro_volcano %>%
    group_by(!!dplyr_i_q90_liv_injury) %>%
    dplyr::select(!!dplyr_i_q90_liv_injury, starts_with(prefix)) %>%
    # get means by groups
    dplyr::summarise_all(.funs =  mean, na.rm = T) %>%
    column_to_rownames(var = quo_name(dplyr_i_q90_liv_injury))
  
  # calculate x-axis for volcano plot
  pro_means <- as_tibble(t(pro_means_messy)) %>% 
    mutate(features = colnames(pro_means_messy)) %>% 
    select(features, everything()) %>% 
    dplyr::mutate(
      fc = Injury/Normal,
      log2fc = log(fc, 2)
      )
  
  # get p-values from t-test, needs to be joined to previous data
  pval_ttest <- df_pro_volcano %>% 
    summarise(
      across(.cols = starts_with(prefix),
             .fns = ~unlist(broom::tidy(t.test(
               .[!!dplyr_i_q90_liv_injury == "Normal"], 
               .[!!dplyr_i_q90_liv_injury == "Injury"]
               ))$p.value)
             )
      ) %>% bind_cols() # make a single vector
  colnames(pval_ttest) <- names(select(df_pro_volcano, starts_with(prefix)))
  
  # join p-values to previous data, format data for volcano plots
  pro_means_test <- left_join(pro_means, by = "features",
    tibble(
      features = names(pval_ttest),
      pval_ttest = as.numeric(pval_ttest)
    )
  ) %>% 
    dplyr::mutate(
      # adjust for multiple comparisons
      pval_fdr = p.adjust(pval_ttest, method = "BH"),
      # negative log 10 for volcano plot
      pval_neglog = -log10(pval_fdr),
      
      # add differential expression
      diffexpr = case_when(
        log2fc < -log2_mag_diff & pval_fdr < alpha ~ 1, # Higher in Normal, Significant
        log2fc < -log2_mag_diff & pval_fdr > alpha ~ 2, # Higher in Normal
        log2fc > -log2_mag_diff & log2fc < log2_mag_diff & pval_fdr > alpha ~ 3, # No Difference
        log2fc > -log2_mag_diff & log2fc < log2_mag_diff & pval_fdr < alpha ~ 4, # Significantly No Difference
        log2fc > log2_mag_diff  & pval_fdr > alpha ~ 5, # Higher in Liver Injury
        log2fc > log2_mag_diff  & pval_fdr < alpha ~ 6, # Significantly Higher in Liver Injury
      ) %>% factor(1:6, c("Significantly Higher in Normal", "Higher in Normal", "No Difference", "Significantly No Difference", "Higher in Liver Injury", "Significantly Higher in Liver Injury")),
      
      # add labels if deferentially expressed
      label_de = ifelse(as.integer(diffexpr) %in% c(1, 2, 5, 6), str_remove(string = features, pattern = trim_label), NA)
    )
  
  # plot
  ggplot(data=pro_means_test, aes(x = log2fc, y = pval_neglog, col = diffexpr)) + 
    geom_point() +
    geom_hline(yintercept = -log10(alpha), col = "gray", linetype = 2) +
    geom_vline(xintercept = c(-log2_mag_diff, log2_mag_diff), col = "gray", linetype = 2) +
    scale_color_manual(name = "Differential\nExpression", values = colors_fct_diffexpr) +
    xlab("Log2 Fold Change") +
    ylab("Significance (-log[P_fdr])") +
    theme(
      legend.position = "bottom",
      legend.title = element_text(size = 10),
      legend.text = element_text(size = 9)
    ) +
    guides(colour = guide_legend(nrow = 2)) +
    labs(caption = paste0("\n", round(mag_diff, 2), " unit difference indicated by vertical bars"))
  
}

label_volcano <- function(df){
  df +
    ggrepel::geom_text_repel(aes(label = label_de), na.rm = TRUE)
}
```

```{r}
# function to remove log scaling from metabolomics features
invert_log <- function(df = df_selection_messy, prefix, base = 2, remove_prefix = " ") {
  df_prefix <- df %>% 
    select(helixid, starts_with("i_"), starts_with(prefix)) %>% 
    mutate(across(starts_with(prefix), ~ base^.))
  
  # remove prefix from column names after reversing the values
  colnames(df_prefix) <- str_remove(colnames(df_prefix), remove_prefix)
  return(df_prefix)
}
# invert_log(prefix = "metser_", remove_prefix = "log.") %>% glimpse
```

## Genetic SNPs{.tabset}

### ALT Sex-Specific{.tabset}

```{r}
p_volc_snps_alt_sex <- plot_volcano(prefix = "rs", i_q90_liv_injury = i_alt_sex, trim_label = "_(A|T|C|G)$") 
```

#### Plot

```{r}
p_volc_snps_alt_sex
```

#### Labels

```{r}
p_volc_snps_alt_sex %>% label_volcano
```

### Enzyme ≥ 90th Percentile{.tabset}

#### Any Enzyme{.tabset}

<span style="color: blue;">rs659366</span> C -> T is a SNP in the [Uncoupling protein-2 (UCP2)](https://pubmed.ncbi.nlm.nih.gov/15970480/). It has been associated with NAFLD as early as 2005.

```{r}
p_volc_snps_any <- plot_volcano(prefix = "rs", i_q90_liv_injury = i_q90_any, trim_label = "_(A|T|C|G)$") 
```

##### Plot

```{r}
p_volc_snps_any
```

##### Labels

```{r}
p_volc_snps_any %>% label_volcano
```

#### ALT{.tabset}

```{r}
p_volc_snps_alt <- plot_volcano(prefix = "rs", i_q90_liv_injury = i_q90_alt, trim_label = "_(A|T|C|G)$") 
```

##### Plot

```{r}
p_volc_snps_alt
```

##### Labels

```{r}
p_volc_snps_alt %>% label_volcano
```

#### AST{.tabset}

```{r}
p_volc_snps_ast <- plot_volcano(prefix = "rs", i_q90_liv_injury = i_q90_ast, trim_label = "_(A|T|C|G)$") 
```

##### Plot

```{r}
p_volc_snps_ast
```

##### Labels

```{r}
p_volc_snps_ast %>% label_volcano
```

#### GGT{.tabset}

```{r}
p_volc_snps_ggt <- plot_volcano(prefix = "rs", i_q90_liv_injury = i_q90_ggt, trim_label = "_(A|T|C|G)$") 
```

##### Plot

```{r}
p_volc_snps_ggt
```

##### Labels

```{r}
p_volc_snps_ggt %>% label_volcano
```

#### CK-18{.tabset}

```{r}
p_volc_snps_ck18 <- plot_volcano(prefix = "rs", i_q90_liv_injury = i_q90_ck18, trim_label = "_(A|T|C|G)$") 
```

##### Plot

```{r}
p_volc_snps_ck18
```

##### Labels

```{r}
p_volc_snps_ck18 %>% label_volcano
```


## Proteomics{.tabset}

```{r}
# remove log2 scaling of metabolomics to get proper volcano plot
df_volc_protein <- df_selection_messy %>% invert_log(prefix = "pro_")
```

### ALT Sex-Specific{.tabset}

```{r}
p_volc_protein_alt_sex <- plot_volcano(df = df_volc_protein, prefix = "pro_", i_q90_liv_injury = i_alt_sex, trim_label = "^pro_")
```

#### Plot

```{r}
p_volc_protein_alt_sex
```

#### Labels

```{r}
p_volc_protein_alt_sex %>% label_volcano
```

### Enzyme ≥ 90th Percentile{.tabset}

#### Any Enzyme{.tabset}

```{r}
p_volc_protein_any <- plot_volcano(df = df_volc_protein, prefix = "pro_", i_q90_liv_injury = i_q90_any, trim_label = "^pro_")
```

##### Plot

```{r}
p_volc_protein_any
```

##### Labels

```{r}
p_volc_protein_any %>% label_volcano
```

#### ALT{.tabset}

```{r}
p_volc_protein_alt <- plot_volcano(df = df_volc_protein, prefix = "pro_", i_q90_liv_injury = i_q90_alt, trim_label = "^pro_")
```

##### Plot

```{r}
p_volc_protein_alt
```

##### Labels

```{r}
p_volc_protein_alt %>% label_volcano
```

#### AST{.tabset}

```{r}
p_volc_protein_ast <- plot_volcano(df = df_volc_protein, prefix = "pro_", i_q90_liv_injury = i_q90_ast, trim_label = "^pro_")
```

##### Plot

```{r}
p_volc_protein_ast
```

##### Labels

```{r}
p_volc_protein_ast %>% label_volcano
```

#### GGT{.tabset}

```{r}
p_volc_protein_ggt <- plot_volcano(df = df_volc_protein, prefix = "pro_", i_q90_liv_injury = i_q90_ggt, trim_label = "^pro_")
```

##### Plot

```{r}
p_volc_protein_ggt
```

##### Labels

```{r}
p_volc_protein_ggt %>% label_volcano
```

#### CK-18{.tabset}

```{r}
p_volc_protein_ck18 <- plot_volcano(df = df_volc_protein, prefix = "pro_", i_q90_liv_injury = i_q90_ck18, trim_label = "^pro_")
```

##### Plot

```{r}
p_volc_protein_ck18
```

##### Labels

```{r}
p_volc_protein_ck18 %>% label_volcano
```

## Metabolomics

```{r}
# remove log2 scaling of metabolomics to get proper volcano plot
df_volc_met_ser <- df_selection_messy %>% invert_log(prefix = "metser_", remove_prefix = "log.")

# remove log2 scaling of metabolomics to get proper volcano plot
df_volc_met_uri <- df_selection_messy %>% invert_log(prefix = "meturi_", remove_prefix = "log.")
```


### Serum{.tabset}

#### ALT Sex-Specific{.tabset}

```{r}
# volcano on non-log transformed data
p_volc_met_ser_alt_sex <- plot_volcano(df = df_volc_met_ser, prefix = "metser_", i_q90_liv_injury = i_alt_sex, trim_label = "^metser_", mag_diff = 1.1)
```

##### Plot

```{r}
p_volc_met_ser_alt_sex
```

##### Labels

```{r}
p_volc_met_ser_alt_sex %>% label_volcano
```

#### Enzyme ≥ 90th Percentile{.tabset}

##### Any Enzyme{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_ser_any <- plot_volcano(df = df_volc_met_ser, prefix = "metser_", i_q90_liv_injury = i_q90_any, trim_label = "^metser_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_ser_any
```

###### Labels

```{r}
p_volc_met_ser_any %>% label_volcano
```

##### ALT{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_ser_alt <- plot_volcano(df = df_volc_met_ser, prefix = "metser_", i_q90_liv_injury = i_q90_alt, trim_label = "^metser_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_ser_alt
```

###### Labels

```{r}
p_volc_met_ser_alt %>% label_volcano
```

##### AST{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_ser_ast <- plot_volcano(df = df_volc_met_ser, prefix = "metser_", i_q90_liv_injury = i_q90_ast, trim_label = "^metser_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_ser_ast
```

###### Labels

```{r}
p_volc_met_ser_ast %>% label_volcano
```

##### GGT{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_ser_ggt <- plot_volcano(df = df_volc_met_ser, prefix = "metser_", i_q90_liv_injury = i_q90_ggt, trim_label = "^metser_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_ser_ggt
```

###### Labels

```{r}
p_volc_met_ser_ggt %>% label_volcano
```


##### CK-18{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_ser_ck18 <- plot_volcano(df = df_volc_met_ser, prefix = "metser_", i_q90_liv_injury = i_q90_ck18, trim_label = "^metser_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_ser_ck18
```

###### Labels

```{r}
p_volc_met_ser_ck18 %>% label_volcano
```


#### Boxplot

```{r}
# elongate data for boxplots
p_df_volc_met_ser <- df_volc_met_ser %>% 
  pivot_longer(names_to = "metabolite", cols = starts_with("metser_"))

# show boxplots
p_df_volc_met_ser %>% 
  ggplot(aes(x = metabolite, y = value)) + 
  geom_boxplot(na.rm = TRUE)
```


### Urine{.tabset}

#### ALT Sex-Specific{.tabset}

```{r}
# volcano on non-log transformed data
p_volc_met_uri_alt_sex <- plot_volcano(df = df_volc_met_uri, prefix = "meturi_", i_q90_liv_injury = i_alt_sex, trim_label = "^meturi_", mag_diff = 1.1)
```

##### Plot

```{r}
p_volc_met_uri_alt_sex
```

##### Labels

```{r}
p_volc_met_uri_alt_sex %>% label_volcano
```

#### Enzyme ≥ 90th Percentile{.tabset}

##### Any Enzyme{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_uri_any <- plot_volcano(df = df_volc_met_uri, prefix = "meturi_", i_q90_liv_injury = i_q90_any, trim_label = "^meturi_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_uri_any
```

###### Labels

```{r}
p_volc_met_uri_any %>% label_volcano
```

##### ALT{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_uri_alt <- plot_volcano(df = df_volc_met_uri, prefix = "meturi_", i_q90_liv_injury = i_q90_alt, trim_label = "^meturi_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_uri_alt
```

###### Labels

```{r}
p_volc_met_uri_alt %>% label_volcano
```

##### AST{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_uri_ast <- plot_volcano(df = df_volc_met_uri, prefix = "meturi_", i_q90_liv_injury = i_q90_ast, trim_label = "^meturi_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_uri_ast
```

###### Labels

```{r}
p_volc_met_uri_ast %>% label_volcano
```

##### GGT{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_uri_ggt <- plot_volcano(df = df_volc_met_uri, prefix = "meturi_", i_q90_liv_injury = i_q90_ggt, trim_label = "^meturi_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_uri_ggt
```

###### Labels

```{r}
p_volc_met_uri_ggt %>% label_volcano
```

##### CK-18{.tabset}
```{r}
# volcano on non-log transformed data
p_volc_met_uri_ck18 <- plot_volcano(df = df_volc_met_uri, prefix = "meturi_", i_q90_liv_injury = i_q90_ck18, trim_label = "^meturi_", mag_diff = 1.1)
```

###### Plot

```{r}
p_volc_met_uri_ck18
```

###### Labels

```{r}
p_volc_met_uri_ck18 %>% label_volcano
```

#### Boxplot

```{r}
# elongate data for boxplots
p_df_volc_met_uri <- df_volc_met_uri %>% 
  pivot_longer(names_to = "metabolite", cols = starts_with("meturi_"))

# show boxplots
p_df_volc_met_uri %>% 
  ggplot(aes(x = metabolite, y = value)) + 
  geom_boxplot(na.rm = TRUE)
```

## Heatmap of Participants by Omics{.tabset}

This heat map illustrates clusters of participants by the hundreds of omics features. All biomarkers have been normalized.

Participants with more than 15% missing omics features are excluded. This is because some people had proteomics measures, but not serum metabolites, leading to whole columns of missing. Missing features are shown in gray.

The top spectrum highlights different features of the participants, to quickly scan for differences by outcome and covariate variables.

The side spectrum illustrates the different classes of omics features (e.g. proteins, serum metabolites, etc.).

```{r}

# function that allows the horizontal spectrum to vary by providing different covariate and color scheme
p_heat_part_omics <- function(strata_var, colors_pal){
  
  # get data of omics features, log2 and then Pareto transformed
  df_heatmap_omics_all_miss <- df_selection_messy %>% 
    select(helixid, starts_with(c("pro_", "metser_", "meturi_")))
  
  # remove rows (participants) with missing at least p = 85% of variables
  df_heatmap_omics_all <- df_heatmap_omics_all_miss[!rowSums(is.na(df_heatmap_omics_all_miss)) >= ncol(df_heatmap_omics_all_miss)*0.85,]
  
  # get horizontal side colors for participant characteristics
  i_participants <- left_join(by = "helixid",
    df_heatmap_omics_all %>% dplyr::select(helixid), 
    df_selection_messy %>% dplyr::select(-starts_with(c("rs", "pro_", "metser_", "meturi_")))
    ) 
  
  # remove helix id for plotting
  df_heatmap_omics_only <- df_heatmap_omics_all %>% select(-helixid)
  
  # get vertical side colors for omics group categories
  i_features <- data.frame(features = colnames(df_heatmap_omics_only)) %>% 
    group_feature_prefix(feature_group = features)
  
  # transpose df for heatmap, remove helixid
  df_heatmap_t <- t(df_heatmap_omics_only)
  
  # row names (omics features)
  rownames(df_heatmap_t) <- clean_feat_labs(rownames(df_heatmap_t))
  
  # plot heatmap
  gplots::heatmap.2(df_heatmap_t, Rowv = T, Colv = T, scale = "row",
                    col = "bluered", na.color = "gray", key = TRUE, trace = "none", 
                    na.rm = TRUE, dendrogram = "both", hclustfun = function(x) hclust(x, method = "ward.D"),
                    RowSideColors = colors_fct_omics_dark[as.character(i_features[["omics_group"]])],
                    ColSideColors = colors_pal[as.character(i_participants[[strata_var]])]
                    )
  
}
```

### ALT Sex-Specific
```{r}
p_heat_part_omics("i_alt_sex", colors_fct_liv_inj)
```

### Enzyme ≥ 90th Percentile{.tabset}

#### Any Injury
```{r}
p_heat_part_omics("i_q90_any", colors_fct_liv_inj)
```

#### ALT
```{r}
p_heat_part_omics("i_q90_alt", colors_fct_liv_inj)
```

#### AST
```{r}
p_heat_part_omics("i_q90_ast", colors_fct_liv_inj)
```

#### GGT
```{r}
p_heat_part_omics("i_q90_ggt", colors_fct_liv_inj)
```

#### CK-18
```{r}
p_heat_part_omics("i_q90_ck18", colors_fct_liv_inj)
```


### Cohort
```{r}
p_heat_part_omics("h_cohort", colors_fct_cohorts_short)
```


# Principal Component Analysis

```{r}
# get data of omics features, log2 and then Pareto transformed
pareto_log2_omics <- df_selection_messy %>% 
  select(helixid, h_cohort, starts_with(c("pro_", "metser_", "meturi_"))) %>% 
  # remove participants with missing observations
  drop_na() %>%   
  # transform proteomics data by log2...
  #  unsure what the transformation was...
  #  but 1.9% of protein values are negative
  # mutate(across(starts_with("pro_"), ~ log(., base = 2))) %>%
  paretoscale(prefix = c("pro_", "metser_", "meturi_"))

# subset to only include omics features
features_only <- select(pareto_log2_omics, -helixid, -h_cohort)

# get liver injury outcome variables to add color categories to heatmap of participants 
i_participants <- left_join(by = "helixid",
  pareto_log2_omics %>% dplyr::select(helixid), 
  df_selection_messy %>% dplyr::select(-starts_with(c("rs", "pro_", "metser_", "meturi_")))
  ) 

# get indicators of feature type to add color categories to heatmap of features
i_features <- data.frame(features = colnames(features_only)) %>% 
  group_feature_prefix(feature_group = features)

# principle components analysis call and output
#  each row represents an observation (participant sample) 
#  each column represents a variable (peak of biologic feature)
pca <- prcomp(features_only, center = F, scale = F)
pcaresults <- summary(pca)
```

`r nrow(pareto_log2_omics)` participants had complete omics data for principal component analysis.

Missing a lot of observations because some samples did not have proteomics measurements.

Attempted to fill in missing with imputation, but not enough non-missing data because it's all-or-nothing for a given omics layer (e.g., proteomics, serum metabolomics, etc.)

Visualization from [STHDA](http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/118-principal-component-analysis-in-r-prcomp-vs-princomp/) blogpost.

## Scree Plot{.tabset}

```{r}
# scree data
scree.data <- as.data.frame(pcaresults$importance)

# cumulative scree data for plotting
scree_cume <- scree.data["Cumulative Proportion", ] %>% 
  t() %>% as.data.frame() %>% rownames_to_column("Principal Component") %>% 
  mutate(
    `Principal Component` = str_remove(`Principal Component`, "^PC") %>% as.integer(),
    thresh_fill = ifelse(`Cumulative Proportion` >= .9, 5,
                         ifelse(`Cumulative Proportion` >= .8, 4, 
                                ifelse(`Cumulative Proportion` >= .7, 3, 
                                       ifelse(`Cumulative Proportion` >= .6, 2, 1)))) %>% factor(1:5)
    )
```

### Cumulative Variance

```{r}
scree_cume %>% 
  filter(`Principal Component` <= 100) %>% 
  ggplot(aes(x = `Principal Component`, y = `Cumulative Proportion`, fill = thresh_fill)) +
  geom_col() +
  scale_fill_manual(values = c("1" = "gray", "2" ="green", "3" ="yellow", "4" = "orange", "5" = "red")) +
  scale_y_continuous(limits = 0:1, breaks = seq(0,1,.2)) +
  scale_x_continuous(breaks = seq(from = 0, to = 100, by = 5)) +
  geom_hline(yintercept = 0.60, linetype = 2, color = "green") +
  geom_hline(yintercept = 0.70, linetype = 2, color = "yellow") +
  geom_hline(yintercept = 0.80, linetype = 2, color = "orange") +
  geom_hline(yintercept = 0.90, linetype = 2, color = "red") +
  theme(legend.position = "none")
```

### Proportion of Variance

```{r}
factoextra::fviz_eig(pca, ggtheme = theme_cowplot(), choice = "variance", ncp = 30)
```

## Loadings 

### Plot{.tabset}

Assess how -omics biologic features load onto first two principle components

```{r}
# loadings for plotting features onto components
loadings.data <- bind_cols(
  features = colnames(features_only), 
  as.data.frame(pcaresults$rotation)[,1:5]
  ) %>% 
  group_feature_prefix(feature_group = features) 

# define cutoff of interest for PCA
threshold_difference <- 0.09

# get features with strong loadings defined by threshold cutoff
significant.loadings <- loadings.data %>% 
  subset(
    PC1 > threshold_difference | PC1 < -threshold_difference | 
    PC2 > threshold_difference | PC2 < -threshold_difference
  )

# highlight features with loadings in different directions
loadings_aes <- loadings.data %>% 
  mutate(pc.change = case_when(
    PC2 >  threshold_difference ~ 4,
    PC2 < -threshold_difference ~ 5, 
    PC1 >  threshold_difference ~ 2,
    PC1 < -threshold_difference ~ 3, 
    abs(PC1) < threshold_difference & abs(PC2) < threshold_difference ~ 1
    ) %>% factor(1:5, c("No change", "+ correlation PC1", "- correlation PC1", "+ correlation PC2", "-  correlation PC2")),
    label_all = clean_feat_labs(features),
    label_sig_load = ifelse(as.integer(pc.change) %in% c(2:5), label_all, NA)
    )
```

```{r}
# plot of principal components
plot_pc_loadings <- loadings_aes %>% 
  ggplot(aes(x = PC1, y = PC2, color = omics_group)) +
  geom_point(aes(shape = pc.change)) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank()
    ) +
  guides(
    colour = guide_legend(nrow = 3),
    shape = guide_legend(nrow = 3)
    ) +
  scale_color_manual(values = colors_fct_omics_dark)
```

#### Labelled Features

```{r}
plot_pc_loadings +
  ggrepel::geom_text_repel(aes(label = label_sig_load), point.padding = NA, na.rm = TRUE)
```

#### Features

```{r}
plot_pc_loadings
```

### Heatmap

Numbers along right side indicate the Principal Components.

```{r}
# only include 3 principle components because only 3 types of features (for now)
df_heat_features <- t(pca$rotation[,1:3])

# remove noisy tags from feature names
chopped_names <- clean_feat_labs(colnames(df_heat_features))

# shorten tags indicating feature types (protein, serum metabolite, etc.)
clean_names_pro <- ifelse(str_detect(chopped_names, "^pro_"), paste0(chopped_names, ".pr"), chopped_names) %>% str_remove("^pro_")
clean_names_ms <- ifelse(str_detect(clean_names_pro, "^metser_"), paste0(clean_names_pro, ".ms"), clean_names_pro) %>% str_remove("^metser_")
clean_names_mu <- ifelse(str_detect(clean_names_ms, "^meturi_"), paste0(clean_names_ms, ".mu"), clean_names_ms) %>% str_remove("^meturi_")

# shorten feature names
clean_names <- clean_names_mu
colnames(df_heat_features) <- clean_names

# shorten principle component names
short_pc_names <- str_remove(rownames(df_heat_features), "PC")
rownames(df_heat_features) <- short_pc_names

# create heat map
gplots::heatmap.2(
  df_heat_features, col = "bluered", key = TRUE, margins = c(7, 5), trace = "none", 
  dendrogram = "column", hclustfun = function(x) hclust(x, method = "ward.D"),
  ColSideColors = colors_fct_omics_dark[as.character(i_features[["omics_group"]])]
  )
```



## Scores

### Plot{.tabset}

Assess how HELIX participants load onto principle components

```{r}
# score data for plotting
score.data <- bind_cols(
  helixid = pareto_log2_omics$helixid,  
  as.data.frame(pcaresults$x)[,1:5]
) %>% 
  left_join(by = "helixid", 
            y = df_selection_messy %>% 
              dplyr::select(helixid, h_cohort, starts_with("i_"), ends_with("_n"))
  )

# plot 
p_aes_scores <- function(i_color, color_pal){
  dplyr_i_color <- enquo(i_color)
  score.data %>% 
    ggplot(aes(PC1, PC2, color = !!dplyr_i_color)) +
    geom_point() +
    stat_ellipse() +
    scale_color_manual(name = "", values = color_pal)
}
```

#### ALT Sex-Specific
```{r}
p_aes_scores(i_alt_sex, colors_fct_liv_inj)
```

#### Enzyme ≥ 90th Percentile{.tabset}

##### Any Enzyme
```{r}
p_aes_scores(i_q90_any, colors_fct_liv_inj)
```

##### ALT
```{r}
p_aes_scores(i_q90_alt, colors_fct_liv_inj)
```

##### AST
```{r}
p_aes_scores(i_q90_ast, colors_fct_liv_inj)
```

##### GGT
```{r}
p_aes_scores(i_q90_ggt, colors_fct_liv_inj)
```

##### CK-18
```{r}
p_aes_scores(i_q90_ck18, colors_fct_liv_inj)
```

#### Cohort

```{r}
p_aes_scores(h_cohort, colors_fct_cohorts_short)
```


### Heatmap{.tabset}

Numbers along right side indicate the Principal Components.

```{r}
# data for heatmap of participants
df_heat_participants <- t(pca$x[,1:3])

# shorten principle component names
short_pc_names <- str_remove(rownames(df_heat_participants), "PC")
rownames(df_heat_participants) <- short_pc_names

p_heat_part_pca <- function(strata_var, color_pal){
  gplots::heatmap.2(df_heat_participants, col = "bluered", key = TRUE, margins = c(7, 5), trace = "none", 
                    dendrogram = "column", hclustfun = function(x) hclust(x, method = "ward.D"),
                    ColSideColors = color_pal[as.character(i_participants[[strata_var]])]
                    )
}
```

#### Enzyme ≥ 90th Percentile
```{r}
p_heat_part_pca("i_alt_sex", colors_fct_liv_inj)
```

#### Enzyme ≥ 90th Percentile{.tabset}

##### Any Enzyme
```{r}
p_heat_part_pca("i_q90_any", colors_fct_liv_inj)
```

##### ALT
```{r}
p_heat_part_pca("i_q90_alt", colors_fct_liv_inj)
```

##### AST
```{r}
p_heat_part_pca("i_q90_ast", colors_fct_liv_inj)
```

##### GGT
```{r}
p_heat_part_pca("i_q90_ggt", colors_fct_liv_inj)
```

##### CK-18
```{r}
p_heat_part_pca("i_q90_ck18", colors_fct_liv_inj)
```

#### Cohort
```{r}
p_heat_part_pca("h_cohort", colors_fct_cohorts_short)
```


# References

